src/ucx_rma_plugin.c

```diff
@@ -26,1239 +23,1277 @@
   }
 
-  return ncclSuccess;
-}
-
-static void nccl_ucx_rma_am_isend_cb(void *request, ucs_status_t status)
-{
-  nccl_ucx_am_request_t *req = (nccl_ucx_am_request_t*)request;
-
-  req->req->done |= NCCL_UCX_RMA_REQUEST_AM_DONE;
-  return;
+  return nccl_rkey->rkey;
 }
 
-static void nccl_ucx_rma_put_isend_cb(void *request, ucs_status_t status, void *data)
-{
-  nccl_ucx_rma_request_t *req = (nccl_ucx_rma_request_t*)data;
-
-  req->done |= NCCL_UCX_RMA_REQUEST_PUT_DONE;
-  return;
-}
+static ncclResult_t nccl_ucp_send(nccl_ucp_comm_t *comm, unsigned short id,
+                                  unsigned i, void *data, int size,
+                                  nccl_ucp_memh_t *mh, void **request) {
+  nccl_ucp_req_t *req;
+  nccl_ucp_rtr_t *rtr;
+  nccl_ucp_atp_t *atp;
+  ucs_status_ptr_t status_ptr;
+  ucp_request_param_t param;
+  ucp_rkey_h rkey;
 
-ncclResult_t nccl_ucx_rma_isend(void *send_comm, void *data, int size, int tag,
-                                void *mhandle, void **request)
-{
-  nccl_ucx_rma_send_comm_t     *comm = (nccl_ucx_rma_send_comm_t*)send_comm;
-  ucx_rma_mhandle_t            *mh   = (ucx_rma_mhandle_t*)mhandle;
-  volatile ucx_rma_send_fifo_t *slot;
-  volatile uint32_t            *ready_ptr;
-  volatile int                 *rkey_id;
-  volatile int                 *rkey_index;
-  nccl_ucx_rma_request_t       *req;
-  ucs_status_ptr_t             st;
-  int                          req_id;
-  ucp_request_param_t          req_param;
-
-  if (comm->super.ready != NCCL_UCX_RMA_SCOMM_READY) {
-    NCCLCHECK(nccl_ucx_rma_send_check(comm));
-    if (comm->super.ready != NCCL_UCX_RMA_SCOMM_READY) {
-      *request = NULL;
-      return ncclSuccess;
-    }
+  *request = NULL;
+  atp      = &comm->local.share.atp[id & NCCL_UCP_RING_MASK];
+  rtr      = &comm->local.share.rtr[id & NCCL_UCP_RING_MASK];
+  req      = &comm->req[comm->req_id & NCCL_UCP_RING_MASK];
+  assert(req->comm == NULL);
+  assert(rtr->avail > 0);
+  assert(rtr->id_start == id);
+
+  rkey = nccl_ucp_rkey_get(comm, rtr->chunk[i].rkey_id);
+  if (rkey == NULL) {
+    return ncclSuccess;
   }
 
-  slot       = comm->fifo + (comm->fifo_head % MAX_REQUESTS);
-  ready_ptr  = &slot->ready;
-  rkey_id    = &slot->rkey_id;
-  rkey_index = &slot->rkey_idx;
-
-  if ((*ready_ptr == 0) ||
-      (comm->rkeys[*rkey_index].id != *rkey_id)) {
-    ucp_worker_progress(comm->super.worker);
-    *request = NULL;
+  param.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
+                       UCP_OP_ATTR_FIELD_USER_DATA | UCP_OP_ATTR_FIELD_MEMH |
+                       UCP_OP_ATTR_FIELD_MEMORY_TYPE;
+  param.cb.send      = nccl_ucp_rdma_isend_callback;
+  param.user_data    = req;
+  param.memh         = mh->ucp_memh;
+  param.memory_type  = mh->mem_type;
+
+  status_ptr =
+      ucp_put_nbx(comm->ucp_ep, data, size, rtr->chunk[i].data, rkey, &param);
+  if ((UCS_PTR_STATUS(status_ptr) != UCS_INPROGRESS) &&
+      (UCS_PTR_STATUS(status_ptr) != UCS_OK)) {
```

Comment:
```suggestion
  if (UCS_PTR_IS_ERR(status_ptr)) {
```
also maybe print error here?

Reply:
1- fixed
2- impossible as it is hot path and even when disabled logs consume cpu